exploration/function-composition-part-1.md

```diff
@@ -0,0 +1,974 @@
+A `FormattedValue` represents something that
+already contains all the information needed
+to either render it as a string,
+or convert it to a "part" according to a hypothetical
+"format to parts" interface.
+
+It would be tempting to use the existing `Formattable`
+type for the input of a function implementation,
+and use the existing `FormattedValue` type for the output.
+However, functions with this signature
+can't compose with each other.
+
+When the following text refers to `Formattable` or
+`FormattedValue`, it should be taken to refer to
+abstract "input" and "output" types.
+
+### Ambiguous examples
+
+Returning to Example Y1, consider two possible models
+of the runtime behavior of function composition.
+
+**Model 1:**
+
+1. Evaluate `$num` to a value and pass it to the `:number` function,
+   along with named options `{"maxFrac": "2"}`
+2. Bind `$x` to the result, which is an object `X`
+   encapsulating the following fields:
+     * The source value, `"0.33333"`
+     * The fully-evaluated options, `{"maxFrac": "2"}`
+     * The formatted result, a `FormattedNumber` object
+       representing the string `"0.33"`
+3. Evaluate `$y` to a value, which is `X`,
+   and pass it to the `:number` function,
+   along with named options `{"maxFrac": "5"}`
+4. Bind `$y` to the result, which is an object `Y`
+   encapsulating the following fields:
+     * The source value, `"0.33333"` (same as `X`'s source value)
+     * The fully-evaluated options, `{"maxFrac": "5"}`
+       (note: the original `maxFrac` option value has been discarded)
+     * The formatted result, a `FormattedNumber` object
+       representing the string `"0.33333"`
+
+then the formatted result is "0.33 0.33333".
+
+**Model 2:**
+
+1. Evaluate `$num` to a value and pass it to the `:number` function,
+   along with named options `{"maxFrac": "2"}`
+2. Bind `$x` to the result, which is:
+     * The formatted result F, a `FormattedNumber` object
+       representing the string `"0.33"`
+3. Evaluate `$y` to a value, which is F,
+   and pass it to the `:number` function,
+   along with named options `{"maxFrac": "5"}`
+4. Inside the `:number` function, check the type
+   of the input, notice that it is already a
+   `FormattedNumber`, and return it as-is
+5. Bind `$y` to the result, which is F.
+
+then the formatted result is "0.33 0.33".
+
+The difference is in step 2: whether the implementation
+of the `number` function returns a value encapsulating
+the various options that were passed in (model 1),
+or only a formatted result (model 2).
+
+In terms of implementation, the result depends on
+what the nature is of the value that is bound to
+a local variable in the environment used in evaluation
+within the message formatter.
+The structure of this value might not be exactly the same
+as the structure that is passed to functions:
+for example, in a lazy implementation, unevaluated thunks
+might be stored in the environment, and then evaluated
+before a function call.
```

Comment:
interesting. that actually sounds like an important point, even if just an implementation detail. because if the thunk is something like `getSystemTimeInMillisNow()`, you want to evaluate and reuse, rather than calling it again & again and getting different user-visible results for the same thing in the same message :-)

Reply:
This is indeed an important point, and not just an implementation detail IMO.

Call-by-need and call-by-name implementations of laziness are observationally equivalent (except when it comes to runtime errors) in a language that doesn't have mutability or side effects.

MF2 is immutable and has no side effects, _unless_ functions have side effects that are observable from messages. `getSystemTimeInMillisNow()` is a good example of a function that does.

It's orthogonal to this design doc, I think, but maybe it should be noted somewhere in the spec that if custom functions with observable side effects exist, then careful thought should be put into lazy implementations.